genai-perf/genai_perf/goodput_calculator/llm_goodput_calculator.py

```diff
@@ -0,0 +1,183 @@
+from genai_perf.metrics.llm_metrics import LLMMetrics
+from genai_perf.metrics.metrics import Metrics
+
+logger = logging.getLogger(__name__)
+
+
+class LLMGoodputCalculator(GoodputCalculator):
+    """
+    A subclass to calculate goodput for LLMs according to LLM-related SLOs.
+    """
+
+    def __init__(
+        self,
+        goodput_constraints: Dict[str, float],
+        metric: Union[LLMMetrics, Metrics],
+        benchmark_duration: float,
+    ) -> None:
+        super().__init__(goodput_constraints, metric, benchmark_duration)
+        self._valid_time_related_names = [
+            item.name for item in metric.request_time_metrics
+        ]
+        self._valid_throughput_related_names = [
+            item.name for item in metric.request_throughput_metrics
+        ]
+        self._valid_metric_names = (
+            self._valid_time_related_names + self._valid_throughput_related_names
+        )
+        self._has_time_target = False
+        self._has_throughput_target = False
+
+        # add slo base name mapping
+        self._slo_base_names["time_to_first_token"] = "time_to_first_tokens"
+        self._slo_base_names["inter_token_latency"] = "inter_token_latencies"
+        self._slo_base_names["output_token_throughput_per_request"] = (
+            "output_token_throughputs_per_request"
+        )
+
+    def _set_valid_slos(self) -> None:
+        """
+        Check users' Service Level Objectives (SLOs) inputs.
+        Set the valid ones while logging the invalid ones.s
+        """
+        invalid_slos = []
+        self._valid_time_related_slos = {}
+        self._valid_throughput_related_slos = {}
+        for slo_name, slo_value in self._goodput_constraints.items():
+            if slo_name in self._valid_metric_names:
+                if slo_name in self._valid_time_related_names:
+                    self._valid_time_related_slos[slo_name] = (
+                        slo_value * self.MS_TO_NS_CONVERSION
+                    )
+                elif slo_name in self._valid_throughput_related_names:
+                    self._valid_throughput_related_slos[slo_name] = slo_value
+            else:
+                invalid_slos.append(slo_name)
+        if self._valid_time_related_slos:
+            self._has_time_target = True
+        if self._valid_throughput_related_slos:
+            self._has_throughput_target = True
+        if invalid_slos:
+            valid_slos_list = ", ".join(self._valid_metric_names)
+            logger.info(
+                f"Invalid SLOs found: {', '.join(invalid_slos)}. "
+                f"The goodput will be -1. Valid SLOs are: {valid_slos_list}."
+            )
+            self._goodput = self.INVALID_GOODPUT
+
+    def _combine_requests_metric_values(self) -> None:
+        """
+        Combine values from the metrics that match with the valid SLOs at a
+        per request level.
+        """
+        if self.goodput == self.INVALID_GOODPUT:
+            return
```

Comment:
what happens if these methods return early? do the rest of the code account for that?

Reply:
Yes, only when there are invalid constraints in users' inputs, goodput will be INVALID_GOODPUT, the rest of code doesn't need to execute.